src/scenario/scenario.c

```diff
@@ -7,36 +7,124 @@
+    state_offsets.random_events = state_offsets.glademp + 16;
+    state_offsets.fishing = state_offsets.random_events + 36;
+    state_offsets.old_filler_3 = state_offsets.fishing + (MAX_FISH_POINTS * 4);
+    state_offsets.invasions_part2 = state_offsets.old_filler_3 + 20;
+    state_offsets.old_filler_4 = state_offsets.invasions_part2 + (MAX_INVASIONS * 1); // 1460
+    state_offsets.rome_wheat = state_offsets.old_filler_4 + 80;
+    state_offsets.allowed_buildings = state_offsets.rome_wheat + 4;
+    state_offsets.win_criteria = state_offsets.allowed_buildings + (MAX_ALLOWED_BUILDINGS * 2); // 1644
+    state_offsets.map_points = state_offsets.win_criteria + 52;
+    state_offsets.invasions_part3 = state_offsets.map_points + 12;
+    state_offsets.misc = state_offsets.invasions_part3 + (MAX_INVASION_POINTS * 4); // 1740
+    state_offsets.custom_name = state_offsets.misc + 51; // 1791
+    state_offsets.end = state_offsets.custom_name + 51; // 1842. Not sure how custon name length was determined.
+}
+
+int scenario_get_state_buffer_size_by_savegame_version(int savegame_version)
+{
+    if (savegame_version <= SAVE_GAME_LAST_UNVERSIONED_SCENARIOS) {
+        return 1720;
+    } else if (savegame_version <= SAVE_GAME_LAST_NO_EXTENDED_REQUESTS) {
+        return 1838;
+    } else {
+        state_offsets_init();
+        return state_offsets.end;
+    }
+}
+
+int scenario_get_state_buffer_size_by_scenario_version(int scenario_version)
+{
+    if (scenario_version <= SCENARIO_LAST_UNVERSIONED) {
+        return 1720;
+    } else if(scenario_version <= SCENARIO_LAST_NO_WAGE_LIMITS) {
+        return 1830;
+    } else if (scenario_version <= SCENARIO_LAST_NO_EXTENDED_REQUESTS) {
+        return 1838;
+    } else {
+        state_offsets_init();
+        return state_offsets.end;
+    }
+}
+
 void scenario_save_state(buffer *buf)
 {
+    state_offsets_init();
+
+    int buf_size = scenario_get_state_buffer_size_by_scenario_version(SCENARIO_CURRENT_VERSION);
+    uint8_t *buf_data = malloc(buf_size);
+    buffer_init(buf, buf_data, buf_size);
+    
+    // size
+    buffer_write_i32(buf, buf_size);
+
+    // start_info
     buffer_write_i16(buf, scenario.start_year);
     buffer_write_i16(buf, 0);
     buffer_write_i16(buf, scenario.empire.id);
     buffer_skip(buf, 8);
 
-    // requests
-    for (int i = 0; i < MAX_REQUESTS; i++) {
-        buffer_write_i16(buf, scenario.requests[i].year);
-    }
-    for (int i = 0; i < MAX_REQUESTS; i++) {
-        buffer_write_i16(buf, scenario.requests[i].resource);
-    }
-    for (int i = 0; i < MAX_REQUESTS; i++) {
-        buffer_write_i16(buf, scenario.requests[i].amount);
-    }
-    for (int i = 0; i < MAX_REQUESTS; i++) {
-        buffer_write_i16(buf, scenario.requests[i].deadline_years);
-    }
+    // Using fillers for where old data was
+    // otherwise offsets become a mess for old versus new saves.
+    // old_filler_1 - requests
+    buffer_skip(buf, 160);
```

Comment:
I understand your reasoning, but I'm not sure about this. Granted it's not a lot of wasted space and it's even compressed.. let me sleep on it 😉

Reply:
I had initially tried getting rid of this, but I ended up getting so many errors during testing that I gave up in the end.
I did do a few refactors after that though, so might be easier to sort out now (mainly since offsets function + struct was added.)

Reply:
I'm leaving this as-is for now. Let me know if I should revisit this.

Reply:
Allright, I bit the bullet and refactored this again.
Filler buffer bytes are now removed and offset function updated to take at least changes from SCENARIO_LAST_NO_EXTENDED_REQUESTS and forward into account.

I do not have old savegames from version where buffer was even more different, so I'm not going to even attempt incorporating the other:
`if (version) {
   buffer size or buffer position
 }`
code changes into the offset function.
I've done enough housekeeping here for now.

I will do thorough testing of this once you are happy with the changes and I have rebased and squashed it on latest, but so far so good.